### PR TITLE
Refactoring the instrumentation component to support multiple versions of Python

### PR DESCRIPTION
**Pull request by BergLucas**
Hi,

After more time than I had estimated, I finally finished making this pull request to add support for multiple versions of Python in Pynguin by refactoring the instrumentation component.

Here is a summary of the changes I made and the bugs I fixed:

## Changes summary

- Refactoring of the instrumentation component architecture to facilitate Pynguin's support for multiple versions of Python
- Add support for Python 3.11, 3.12 and 3.13
- Refactoring of the control-flow graph to improve code reuse in other modules
- Refactoring of the tracer to reduce data transfer and improve the performance of subprocess execution.

## Bugs

- Fixed the fact that the `object_addresses` in the subject properties were not reset after each test execution
- Fixed the use of `CODE_OBJECT_ID_KEY` in code object docstrings, which was breaking some modules that performed operations on their docstrings
- Fixed some edge cases in the dynamic slicer

## Explanation

Initially, I hadn't planned on making so many changes, but the code was more coupled than I thought, so I didn't have the choice.

### Initial analysis and refactoring of the CFG

First, I analysed the [InstrumentationAdapter](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/instrumentation.py#L121)s and I noticed that they were directly using `BasicBlock`s when they could have used [ProgramNode](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/analyses/controlflow.py#L37)s to have a class in which to define helper functions. However, this wasn't easily possible because there was only one class for basic block nodes and artificial nodes, so it would have been necessary to use assertions everywhere, like there [[1]](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/instrumentation.py#L301)  [[2]](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/instrumentation.py#L344). Given that there are many places where helper functions located in nodes could be useful, I decided that the first thing to do was to refactor the nodes and the CFG so that it would be more convenient to use them in the adapters, especially since they had been created primarily for that purpose. Therefore, I divided the nodes into `BasicBlockNode`s and `ArtificialNode`s, and I extracted as many helper functions as possible to maximise code reuse. This also allowed me to identify some issues with the CFG design. In particular, some fields were present in the nodes, such as an [offset](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/analyses/controlflow.py#L72) that was only used in the [dynamic slicer](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/slicer/executionflowbuilder.py#L73) and could be avoided, or a [predicate_id](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/analyses/controlflow.py#L108) that could also be retrieved from the `SubjectProperties` and could also be avoided. Thus, I refactored the code to remove them and then moved the module to the `instrumentation` package because it wasn't very pertinent in the `analyses` package anymore.

### Refactoring of the tracer and addition of an abstraction layer on top of the bytecode

To instrument the code, we use the [bytecode](https://pypi.org/project/bytecode/) library, which greatly simplifies the process. However, it is not abstract enough between Python versions for what we were doing, i.e., adding calls to our tracer functions. Therefore, I examined how I could abstract the operations we performed during instrumentation into classes that could be reused between Python versions and converted differently to account for changes in the Python bytecode that manages function calls. At first, I thought I would convert all the operations exactly as they were performed, but I noticed that in reality, we sometimes called built-in functions such as `id` in addition to the tracer. This complicated the process, even if we could call these functions directly in the tracer if we passed it the right data. Thus, to keep the abstraction layer as simple as possible, I refactored the tracer to call additional functions in it. It was during this refactoring that I found the bug concerning `object_addresses` by comparing the tracer code with the one of `pychecco`. The bug was that some addresses could persist between multiple executions of tests, although this was not the case in the `pychecco` code on which this module was based. At the same time, I thought it was the right moment to go a little further and reverse the relationship between the [ExecutionTracer](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/tracer.py#L998) and the [SubjectProperties](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/tracer.py#L342), because I noticed that the `SubjectProperties` were no longer necessary in the `ExecutionTracer` after this change, and it was therefore possible to fix the problem of sending excessive data during subprocess execution.

### Refactoring of the `InstrumentationTransformer` and `InstrumentationAdapter`s

Once the abstraction layer was established, it had to be used in such a way that maximised code reuse between versions. So I created a set of specialized instrumentation sub-adapters for each type of instrumentation and designed them to have a good combination of inheritance and composition.  I also made sure that the [InstrumentationTransformer](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/instrumentation.py#L223) could be the same regardless of the Python version, and this is when I found the bug concerning [CODE_OBJECT_ID_KEY](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/instrumentation.py#L37). I was already aware of the problem because I had already encountered modules that returned an error with this name, but I never looked into how to solve it. After analysing this constant, I found that the `code_object_id`s were stored in JSON format in the docstrings of the code objects with this constant and then used to retrieve them from another location. However, I noticed that this information could be retrieved from the `SubjectProperties`, so I was able to remove it. Then I implemented the instrumentation adapters specific to each version of Python.

### Refactoring of the `DynamicSlicer`

Once all of this was implemented, I thought it was almost done, but that wasn't the case. I had never really used the [DynamicSlicer](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/slicer/dynamicslicer.py#L132), nor had I looked closely at what the code was doing, but when I ran the tests, I noticed that almost all of the tests of the `DynamicSlicer` were failing. I tried to debug it, but the code was very complex, there were no logs, and the tests were difficult to understand. Thus, I started a major log/refactoring phase, and it was during this phase that I noticed that this component was also highly dependent on Python bytecode and did not support some edge cases because they did not exist in Python 3.10. It was also while refactoring this code that I discovered the purpose of the `offset` attribute of CFG nodes. It was used to [retrieve information about decompiled code related to an executed instruction](https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/slicer/executionflowbuilder.py#L76). However, starting with Python 3.11, a cache was implemented in the bytecode, which skewed all the offsets, so I had to find a solution. The solution was that the information about the decompiled code was already almost available using the `bytecode` library, so I was able to replace this offset with an instruction index, and it worked.

## Conclusion

In conclusion, I was able to refactor Pynguin's instrumentation component to support multiple versions of Python, even though this required a large number of changes. I would have liked to decrease the size of the PR by dividing it into smaller PRs, but unfortunately, apart from moving the few files that I moved in this PR, everything is so tightly coupled that it is not possible to do much more.

What do you think? Don't hesitate to share your comments.

Have a nice day!

Fix: #84 
Related to: #103, #109

### CODE REVIEW

**Review by LuKrO2011 — CHANGES_REQUESTED** [review]
Hi,
first of all: Thank you so much for your contribution. This is a hudge advancement for Pynguin.
I started reviewing the changes today. As this is a lot, I will request changes step by step.

File: src/pynguin/instrumentation/tracer.py [hunk 1]
```diff
@@ -338,26 +358,192 @@ def __eq__(self, other: object) -> bool:
...
+
+DEFAULT_CODE_OBJECT_ID = 0
+
+
 @dataclass
 class SubjectProperties:
     """Contains properties about the subject under test."""
 
+    # The instrumentation tracer that is used to trace the execution
+    instrumentation_tracer: InstrumentationExecutionTracer = field(
```
[the remark this answers is not in the document]
Comment (BergLucas) [reply src/pynguin/instrumentation/tracer.py:439]: It already existed before, so I didn't create it: https://github.com/se2p/pynguin/blob/1e4bcffa99267fed9995c56c0c98d39cf675f653/src/pynguin/instrumentation/tracer.py#L342

However, previously, it was the `ExecutionTracer` class that contained a `SubjectProperties`, and now it is the `SubjectProperties` class that contains an `ExecutionTracer`. From what I have seen, the `SubjectProperties` are created once at the start of Pynguin during the module instrumentation. Then, they remain unchanged and are used to calculate metrics and give meaning to what the `InstrumentationExecutionTracer` tracks.

I had considered separating the two, but I realised that to do this properly, it would require refactoring some methods of the observers, so I decided to limit myself to this inversion for now and that we could refactor it better in another PR. Technically, these changes are already a little outside the scope I initially intended for this PR, as I thought I needed to modify the order to obtain some data in a cleaner format in certain places. However, perhaps I could have approached it differently without making the changes in this PR.

File: src/pynguin/instrumentation/controlflow.py [hunk 2]
```diff
@@ -0,0 +1,931 @@
...
+            The raw control-flow graph from the code object
+        """
+        return self._bytecode_cfg
+
+    @staticmethod
+    def _split_try_begin_blocks(blocks: ControlFlowGraph) -> None:
+        for block in blocks:
+            for i, instr in enumerate(block):
+                if isinstance(instr, TryBegin):
+                    # We need to handle the next block manually here because there is a
```
[the remark this answers is not in the document]
Comment (BergLucas) [reply src/pynguin/instrumentation/controlflow.py:568]: I kind of forgot about this haha, I'll look into doing a PR and fix it.

File: src/pynguin/instrumentation/controlflow.py [hunk 3]
```diff
@@ -0,0 +1,931 @@
...
+
+
+class ControlDependenceGraph(ProgramGraph):
+    """Implements a control-dependence graph."""
+
+    @staticmethod
+    def _create_augmented_graph(cfg: CFG) -> ProgramGraph:
+        augmented_graph = ProgramGraph(cfg.graph.copy())
+
+        augmented_graph.add_node(ArtificialNode.AUGMENTED_ENTRY)
```
[the remark this answers is not in the document]
Comment (BergLucas) [reply src/pynguin/instrumentation/controlflow.py:746]: I also considered removing it when I refactored the CDG because I thought there was a bug in it. However, when I read the article explaining the algorithm, I realised that it was something required by the algorithm and that it didn't work without it. https://dl.acm.org/doi/pdf/10.1145/24039.24041 (at the top of page 6 of the PDF)

File: src/pynguin/instrumentation/controlflow.py [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+        """Provides the instructions of the basic block.
+
+        Returns:
+            The instructions of the basic block
+        """
+        for instr in self._basic_block:
+            if isinstance(instr, Instr):
+                yield instr
+
+    def get_instruction(self, index: int) -> Instr:
```
[the remark this answers is not in the document]
Comment (BergLucas) [reply src/pynguin/instrumentation/controlflow.py]: At the beginning, I used it a lot because it allowed me to avoid putting assertions for trivial things in many places, but now it's rarely used, so we can make it private and use the other variant.

**Comment by BergLucas** [comment]
I also had the idea of adding a few lines of configuration so that we could use [tox](https://tox.wiki/en/4.28.4/) and run tests and linters in all supported versions very easily using the `tox` command. If you don't want to use this tool, I can remove the few lines of configuration. If you want to use it in CI, you can do it directly as you have write access to the branch and I don't know if you need to change anything in your internal CI.

File: src/pynguin/instrumentation/controlflow.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LuKrO2011) [reply src/pynguin/instrumentation/controlflow.py:745]: If I understand this correctly, it is required so that the resulting CDG is fully connected. Otherwise, in the example, node 7 would not be connected, because it is not control-dependent on any node but START, yet START and STOP are removed during CDG construction).
I suggest explaining this in a method comment for this method:
```python
"""Augments a CFG (control flow graph) with an additional entry node.

When constructing the CDG (control dependence graph) the entry and exit nodes of the CFG are removed. If the first node of the resulting CDG graph (1) would be used as root node of the CDG, nodes that are only dependent on method entry (7) would not be connected to the graph at all. Thus, we instead need to add an artifical root node: the entry node.

Reference:
Jeanne Ferrante, Karl J. Ottenstein, and Joe D. Warren. 1987. The program dependence graph and its use in optimization. ACM Trans. Program. Lang. Syst. 9, 3 (July 1987), 319–349. https://doi.org/10.1145/24039.24041
See Figure 1 for the referenced example.

Args:
    cfg: The control flow graph
    
Returns:
    A control flow graph with an augmented entry node
"""
```

File: src/pynguin/instrumentation/tracer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LuKrO2011) [reply src/pynguin/instrumentation/tracer.py:439]: I see. I would also prefer not to change this again in this PR as it is already quite large. Let's just add a comment if you agree:
```
TODO(lk): SubjectProperties and ExecutionTracer should be separated
```

File: src/pynguin/instrumentation/version/python3_12.py [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+    match opname[opcode]:
+        case "POP_JUMP_IF_TRUE" | "POP_JUMP_IF_NOT_NONE":
+            return True
+        case "POP_JUMP_IF_FALSE" | "POP_JUMP_IF_NONE" | "FOR_ITER":
+            return False
+        case _:
+            return None
+
+
+RETURN_NONE_SIZE = 1
```
Comment (LuKrO2011) [inline src/pynguin/instrumentation/version/python3_12.py:209]: Constants before methods

**Review by LuKrO2011 — CHANGES_REQUESTED** [review]
I reviewed the `instrumentation.version` package.

File: src/pynguin/instrumentation/version/python3_10.py [hunk 6]
```diff
@@ -0,0 +1,1866 @@
...
+            | "SET_ADD"
+            | "GEN_START"
+        ):
+            return StackEffects(1, 0)
+        case "DELETE_SUBSCR" | "STORE_ATTR" | "MAP_ADD" | "JUMP_IF_NOT_EXC_MATCH":
+            return StackEffects(2, 0)
+        case "STORE_SUBSCR" | "POP_EXCEPT" | "RERAISE":
+            return StackEffects(3, 0)
+        case "END_ASYNC_FOR":
+            return StackEffects(7, 0)  # TODO conditional 0 pops?
```
[the remark this answers is not in the document]
Comment (BergLucas) [reply src/pynguin/instrumentation/version/python3_10.py]: I think so. I didn't find any argument for the stack effects that would do more than 0 pops.

File: src/pynguin/slicer/dynamicslicer.py [hunk 7]
```diff
@@ -166,517 +179,548 @@ def slice(  # noqa: C901
         """
         slc = self._setup_slicing_configuration(slicing_criterion, trace)
 
+        self._logger.debug("========================= [START] =========================")
```
[the remark this answers is not in the document]
Comment (BergLucas) [reply src/pynguin/slicer/dynamicslicer.py:182]: I know it's pretty ugly, but I had no choice because without these well-delimited logs, it was almost impossible to debug the dynamic slicer. Usually, what I tried to do to fix the failing tests was to compare the behaviour of the previous Python version with the one of the new version. However, the logs can be several thousand lines long for the dynamic slicer, so they are very difficult to read. In addition, the slightest problem in one place will affect the rest of the slicer execution. Therefore, I tried to highlight each step of the algorithm using `===` and to uppercase important data so as not to confuse it with function names, etc.

I can remove the `===` and uppercase letters if you want, but in my opinion, that could make it harder to add support for new Python versions in the future, for a feature that is not that essential in my opinion.

Here is an example of the slicer logs for the smallest test:
```log
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:223 STACK PUSHING (IN SLICE): [1 time(s)] <UniqueInstruction RETURN_VALUE arg=0 lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:324 IN SLICE: <UniqueInstruction RETURN_VALUE arg=0 lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:182 ========================= [START] =========================
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:183       <UniqueInstruction RETURN_VALUE arg=0 lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:184 -----------------------------------------------------------
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:342 > INSTRUCTIONS IN SLICE NUMBER: 1
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:378 > LAST FRAME BLOCK INSTRUCTIONS NUMBER: 1 (40 frames)
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:208 ========================= [POSITION 1] =========================
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:217      <UniqueInstruction LOAD_FAST arg=result lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:218 ----------------------------------------------------------------
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:169 STACK POPPING: [1 time(s)]
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:184 POP (IN SLICE): <UniqueInstruction RETURN_VALUE arg=0 lineno=88>
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:190 IMPLICIT DEPENDENCY (IN SLICE): <UniqueInstruction RETURN_VALUE arg=0 lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:269 IMPLICIT DATA DEPENDENCY (STACK): <UniqueInstruction LOAD_FAST arg=result lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:277 CRITERION IN SLICE (IMPLICIT DATA DEPENDENCY): <UniqueInstruction LOAD_FAST arg=result lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:431 IN SLICE: <UniqueInstruction LOAD_FAST arg=result lineno=88>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:757 FAST VARIABLE USE: 'result'
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:342 > INSTRUCTIONS IN SLICE NUMBER: 2
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:349 > VARIABLE LOCAL USES: {'result'}
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:378 > LAST FRAME BLOCK INSTRUCTIONS NUMBER: 0 (40 frames)
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:208 ========================= [POSITION 0] =========================
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:217      <UniqueInstruction STORE_FAST arg=result lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:218 ----------------------------------------------------------------
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:586 EXPLICIT DATA DEPENDENCY (VARIABLE): 'result'
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:243 CRITERION IN SLICE (EXPLICIT DATA DEPENDENCY): <UniqueInstruction STORE_FAST arg=result lineno=87>
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:223 STACK PUSHING (IN SLICE): [1 time(s)] <UniqueInstruction STORE_FAST arg=result lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:431 IN SLICE: <UniqueInstruction STORE_FAST arg=result lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:342 > INSTRUCTIONS IN SLICE NUMBER: 3
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:378 > LAST FRAME BLOCK INSTRUCTIONS NUMBER: 1 (40 frames)
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:213 ========================= [NOT TRACED] =========================
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:217      <UniqueInstruction LOAD_CONST arg=1 lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:218 ----------------------------------------------------------------
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:169 STACK POPPING: [1 time(s)]
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:184 POP (IN SLICE): <UniqueInstruction STORE_FAST arg=result lineno=87>
DEBUG    pynguin.slicer.stack.stacksimulation:stacksimulation.py:190 IMPLICIT DEPENDENCY (IN SLICE): <UniqueInstruction STORE_FAST arg=result lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:269 IMPLICIT DATA DEPENDENCY (STACK): <UniqueInstruction LOAD_CONST arg=1 lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:277 CRITERION IN SLICE (IMPLICIT DATA DEPENDENCY): <UniqueInstruction LOAD_CONST arg=1 lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:431 IN SLICE: <UniqueInstruction LOAD_CONST arg=1 lineno=87>
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:342 > INSTRUCTIONS IN SLICE NUMBER: 4
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:378 > LAST FRAME BLOCK INSTRUCTIONS NUMBER: 0 (40 frames)
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:302 Found 4 instructions in the slice.
DEBUG    pynguin.slicer.dynamicslicer:dynamicslicer.py:303 ========================= [END] =========================
```

**Comment by LuKrO2011** [comment]
Okay, that is fine then.

File: src/pynguin/slicer/dynamicslicer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (LuKrO2011) [reply src/pynguin/slicer/dynamicslicer.py:182]: As it is only a `.debug` log, It's supposed to help with debugging, and I agree that this format is helpful. Thus, let's leave it as it is.
